key.core/src/test/java/de/uka/ilkd/key/speclang/SetStatementTest.java

```diff
@@ -0,0 +1,116 @@
+
+    /**
+     * JavaInfo containing information about the available datatypes and methods
+     */
+    private JavaInfo javaInfo;
+
+    /**
+     * services needed for translation
+     */
+    private Services services;
+
+    /**
+     * service for JML translation
+     */
+    private JmlIO jmlIO;
+
+    /**
+     * context information needed for JmlIO/parser
+     */
+    private KeYJavaType testClassType;
+
+    /**
+     * Creates the JavaInfo, Services, and JmlIO.
+     */
+    @BeforeEach
+    public synchronized void setUp() {
+        if (javaInfo == null) {
+            javaInfo =
+                new HelperClassForTests().parse(new File(TEST_FILE)).getFirstProof().getJavaInfo();
+            services = javaInfo.getServices();
+            testClassType = javaInfo.getKeYJavaType("testPackage.TestClass");
+        }
+        var selfVar = new LocationVariable(new ProgramElementName("self"), testClassType);
+
+        var normalLocal = new LocationVariable(new ProgramElementName("normalLocal"),
+            javaInfo.getKeYJavaType(PrimitiveType.JAVA_INT));
+        var ghostLocal = new LocationVariable(new ProgramElementName("ghostLocal"),
+            javaInfo.getKeYJavaType(PrimitiveType.JAVA_INT), true, false);
+
+        jmlIO = new JmlIO(services)
+                .context(Context.inClass(testClassType, false, services.getTermBuilder()))
+                .selfVar(selfVar)
+                .parameters(ImmutableSLList.<ProgramVariable>nil().append(ghostLocal, normalLocal));
+    }
+
+    @Test
+    public void test() {
+        testError("//@ set this.ghostArray[0] = 0;", "write to array");
+        testError("//@ set this.normalField = 0;", "write to normal field");
+        testNoError("//@ set this.ghostArray[0].ghostField = 0;", "write ghost field");
+        testNoError("//@ set this.ghostField = 0;", "write to ghost field");
+        testNoError("//@ set ghostLocal = 0;", "write to ghost local");
+        testError("//@ set normalLocal = 0;", "write to normal local");
+    }
+
+    private void testError(String statement, String reason) {
+        assertNotNull(parseAndCheck(statement),
+            "'" + statement + "' should produce an error: " + reason);
+    }
+
+    private void testNoError(String statement, String reason) {
+        assertNull(parseAndCheck(statement),
+            "'" + statement + "' should not produce an error: " + reason);
+    }
+
+    private String parseAndCheck(String statementText) {
+        JMLSpecFactory jsf = new JMLSpecFactory(services);
+        ImmutableList<TextualJMLConstruct> constructs =
+            new PreParser().parseMethodLevel(statementText, null, Position.newOneBased(1, 1));
+        assertEquals(constructs.size(), 1);
+        assertInstanceOf(TextualJMLSetStatement.class, constructs.head());
+        var statement = (TextualJMLSetStatement) constructs.head();
+        var context = statement.getAssignment();
+        var assignee = jmlIO.translateTerm(context.assignee);
```

Comment:
type? pls make explicit

Reply:
It's Term and I think that's obvious from the context.